src/node_api_embedding.cc

```diff
@@ -0,0 +1,765 @@
+        platform,
+        &errors,
+        env_options->args_,
+        env_options->exec_args_,
+        env_options->snapshot_config_);
+  } else {
+    env_setup = node::CommonEnvironmentSetup::Create(
+        platform, &errors, env_options->args_, env_options->exec_args_, flags);
+  }
+
+  if (error_handler != nullptr && !errors.empty()) {
+    v8impl::CStringArray error_arr(errors);
+    error_handler(error_handler_data, error_arr.c_strs(), error_arr.size());
+  }
+
+  if (env_setup == nullptr) {
+    return napi_generic_failure;
+  }
+
+  std::string filename =
+      env_options->args_.size() > 1 ? env_options->args_[1] : "<internal>";
+  node::CommonEnvironmentSetup* env_setup_ptr = env_setup.get();
+
+  v8impl::IsolateLocker isolate_locker(env_setup_ptr);
+  std::unique_ptr<v8impl::EmbeddedEnvironment> embedded_env =
+      std::make_unique<v8impl::EmbeddedEnvironment>(std::move(env_options),
+                                                    std::move(env_setup),
+                                                    env_setup_ptr->context(),
+                                                    filename,
+                                                    api_version);
+  embedded_env->node_env()->AddCleanupHook(
+      [](void* arg) { static_cast<napi_env>(arg)->Unref(); },
+      static_cast<void*>(embedded_env.get()));
+  *result = embedded_env.get();
+
+  node::Environment* node_env = env_setup_ptr->env();
+
+  // TODO(vmoroz): If we return an error here, then it is not clear if the
+  // environment must be deleted after that or not.
+  v8::MaybeLocal<v8::Value> ret =
+      embedded_env->snapshot()
+          ? node::LoadEnvironment(node_env, node::StartExecutionCallback{})
+          : node::LoadEnvironment(node_env,
+                                  std::string_view(main_script),
+                                  embedded_env->options().preload_cb_);
+
+  embedded_env.release();
+
+  if (ret.IsEmpty()) return napi_pending_exception;
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL node_api_delete_env(napi_env env, int* exit_code) {
+  CHECK_ENV(env);
+  v8impl::EmbeddedEnvironment* embedded_env =
+      v8impl::EmbeddedEnvironment::FromNapiEnv(env);
+
+  if (embedded_env->IsScopeOpened()) return napi_generic_failure;
+
+  {
+    v8impl::IsolateLocker isolate_locker(embedded_env->env_setup());
+
+    int ret = node::SpinEventLoop(embedded_env->node_env()).FromMaybe(1);
+    if (exit_code != nullptr) *exit_code = ret;
+  }
+
+  std::unique_ptr<node::CommonEnvironmentSetup> env_setup =
+      embedded_env->ResetEnvSetup();
+
+  if (embedded_env->create_snapshot()) {
+    node::EmbedderSnapshotData::Pointer snapshot = env_setup->CreateSnapshot();
+    assert(snapshot);
+    embedded_env->create_snapshot()(snapshot.get());
+  }
+
+  node::Stop(embedded_env->node_env());
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL node_api_open_env_scope(napi_env env) {
```

Comment:
I'm wondering if all functions which are only to be called as part of embedding versus in an add-on implementation should have some extra bit in the name. For exampe in this method `node_api_embed_open_env_scope

Reply:
This open/close scope API gives us a lot of flexibility, but it is difficult to use and like you said it is quite confusing.
I am currently considering to replace it with a function that receives a lambda (c function + void state), and then the `napi_env` will be available only for that lambda. Other APIs will change from using `napi_env` to something like `node_embedding_env` or `node_embedded_env`.